File: python/problems/leetcode/0148_sort_list/solution.py
"""LeetCode 148: Sort List."""

from __future__ import annotations

from typing import Optional


class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val = val
        self.next = next


class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        slow, fast = head, head.next
        while fast is not None and fast.next is not None:
            slow = slow.next  # type: ignore[assignment]
            fast = fast.next.next
        assert slow is not None
        mid = slow.next
        slow.next = None
        left = self.sortList(head)
        right = self.sortList(mid)
        return self._merge(left, right)

    @staticmethod
    def _merge(
        a: Optional[ListNode], b: Optional[ListNode]
    ) -> Optional[ListNode]:
        dummy = ListNode()
        cur = dummy
        while a is not None and b is not None:
            if a.val <= b.val:
                cur.next = a
                a = a.next
            else:
                cur.next = b
                b = b.next
            cur = cur.next
        cur.next = a if a is not None else b
        return dummy.next


def from_list(vs: list[int]) -> Optional[ListNode]:
    dummy = ListNode()
    cur = dummy
    for v in vs:
        cur.next = ListNode(v)
        cur = cur.next
    return dummy.next


def to_list(h: Optional[ListNode]) -> list[int]:
    out: list[int] = []
    while h is not None:
        out.append(h.val)
        h = h.next
    return out


if __name__ == "__main__":
    assert to_list(Solution().sortList(from_list([4, 2, 1, 3]))) == [1, 2, 3, 4]
    assert to_list(Solution().sortList(from_list([-1, 5, 3, 4, 0]))) == [
        -1, 0, 3, 4, 5,
    ]
    assert Solution().sortList(None) is None
    print("0148 OK")

File: cpp/problems/leetcode/0148_sort_list/solution.cpp
// LeetCode 148
#include <alg_std.hpp>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* n) : val(x), next(n) {}
};

class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (!head || !head->next) return head;
        ListNode *slow = head, *fast = head->next;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* mid = slow->next;
        slow->next = nullptr;
        return merge(sortList(head), sortList(mid));
    }

private:
    ListNode* merge(ListNode* a, ListNode* b) {
        ListNode dummy;
        ListNode* cur = &dummy;
        while (a && b) {
            if (a->val <= b->val) {
                cur->next = a;
                a = a->next;
            } else {
                cur->next = b;
                b = b->next;
            }
            cur = cur->next;
        }
        cur->next = a ? a : b;
        return dummy.next;
    }
};

static ListNode* fromList(vector<int> vs) {
    ListNode dummy;
    ListNode* cur = &dummy;
    for (int v : vs) {
        cur->next = new ListNode(v);
        cur = cur->next;
    }
    return dummy.next;
}

static vector<int> toList(ListNode* h) {
    vector<int> out;
    while (h) {
        out.push_back(h->val);
        h = h->next;
    }
    return out;
}

int main() {
    assert((toList(Solution().sortList(fromList({4, 2, 1, 3}))) ==
            vector<int>{1, 2, 3, 4}));
    assert((toList(Solution().sortList(fromList({-1, 5, 3, 4, 0}))) ==
            vector<int>{-1, 0, 3, 4, 5}));
    assert(Solution().sortList(nullptr) == nullptr);
    cout << "0148 OK" << endl;
    return 0;
}
